react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+
+An example of a `Cart` component:
+
+~~~jsx
+function Cart({ products }) {
+  const totalPrice = products.reduce(
+    (total, product) => total + product.price * product.quantity,
+    0
+  );
+
+  return (
+    <div>
+      {/* Some other content in the cart */}
+      {/* Products to display */}
+      <p>
+        Total Price: <strong>${totalPrice}</strong>
+      </p>
+      {/* Some button to checkout */}
+    </div>
+  );
+}
+~~~
+
+In our `Cart` component, we have the total price of the products calculated directly inside the component. Every time the component is rendered or updated, the calculation is performed from scratch! That doesn't sound good... What if the user has added many products to the cart? Then it will lead to a sluggish user experience.
+
+The `reduce` method iterates over each product and performs multiplication and addition for every item in the cart. This operation becomes *increasingly* time-consuming as the number of products increases.
+
+Now imagine a user who frequently opens/closes the cart. Every time the drawer is opened, the `Cart` component is rendered, executing everything inside the component. This results in unnecessary recomputations of the same value even if the cart's content hasn't changed.
+
+Let's see how we can use `useMemo` to address this:
+
+~~~jsx
+import { useMemo } from "react"; // import the `useMemo` hook in the react module
+
+function Cart({ products }) {
+  const totalPrice = useMemo(() => {
+    return products.reduce(
+      (total, product) => total + product.price * product.quantity,
+      0
+    );
+  }, [products]);
+
+  return (
+    <div>
+      {/* Some other content in the cart */}
+      {/* Products to display */}
+      <p>
+        Total Price: <strong>${totalPrice}</strong>
+      </p>
+      {/* Some button to checkout */}
+    </div>
+  );
+}
+~~~
+
+In the example above, we can easily memoize the calculated value by wrapping it in a `useMemo`, as the syntax is pretty much the same as `useEffect` and almost works the same. Where `useMemo` will also *execute* the callback on mount, and on subsequent re-renders, it will only *re-execute* the callback whenever one of the dependencies *changes*. In our case, whenever the `products` prop changes.
+
+This way, whenever a user opens/closes the cart multiple times, it will not recalculate the `totalPrice` and use the cached value as long as`products` did not change.
+
+You can also do an implicit return if you prefer:
+
+~~~jsx
+const totalPrice = useMemo(
+  () =>
+    products.reduce(
+      (total, product) => total + product.price * product.quantity,
+      0
+    ),
+  []
+);
+~~~
```

Comment:
I'm not sure if we want this example. The learner could probably figure out implicit return in `useMemo` themselves.
```suggestion
```